File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n=height.length;
        int l=0,r=n-1;
        int r_max=0,l_max=0,sum=0;
        while(l<r)
        {
            if(height[l]<height[r])
            {
                l_max=Math.max(l_max,height[l]);
                sum+=l_max-height[l];
                l++;
            }
            else
            {
                r_max=Math.max(r_max,height[r]);
                sum+=r_max-height[r];
                r--;
            }
        }
        return sum;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int l = 0, r = n - 1;
        int left_max = 0, right_max = 0;
        int count = 0;

        while (l < r) {
            if (height[l] < height[r]) {
                if (height[l] >= left_max) {
                    left_max = height[l];
                } else {
                    count += left_max - height[l];
                }
                l++;
            } else {
                if (height[r] >= right_max) {
                    right_max = height[r];
                } else {
                    count += right_max - height[r];
                }
                r--;
            }
        }

        return count;
    }
};
